sklearn/model_selection/_split.py

```diff
@@ -456,7 +456,7 @@ class GroupKFold(_BaseKFold):
     distinct groups has to be at least equal to the number of folds).
 
     The folds are approximately balanced in the sense that the number of
-    distinct groups is approximately the same in each fold.
+    samples is approximately the same in each fold.
```

Comment:
I don't think this is true, is it?

Reply:
Yes, it is. The code in question is [this function](https://github.com/scikit-learn/scikit-learn/blob/95d4f0841d57e8b5f6b2a570312e9d832e69debc/sklearn/model_selection/_split.py#L502). On [Line 516](https://github.com/scikit-learn/scikit-learn/blob/95d4f0841d57e8b5f6b2a570312e9d832e69debc/sklearn/model_selection/_split.py#L516) the number of samples per group `n_samples_per_group` is computed by which the groups are weighted. The groups with the most samples are distributed first. Then in the for loop starting on [Line 529](https://github.com/scikit-learn/scikit-learn/blob/95d4f0841d57e8b5f6b2a570312e9d832e69debc/sklearn/model_selection/_split.py#L529) the fold with the least samples gets assigned the next group and the sample count of the fold gets updated accordingly:
```
        # Distribute samples by adding the largest weight to the lightest fold
        for group_index, weight in enumerate(n_samples_per_group):
            lightest_fold = np.argmin(n_samples_per_fold)
            n_samples_per_fold[lightest_fold] += weight
            group_to_fold[indices[group_index]] = lightest_fold
```
The number of groups per fold is irrelevant here and not considered as a weight to distribute the groups across the splits. Instead, the number of samples per fold is what is balanced.

See this example based on the function in question:
```
def _iter_test_indices(groups, n_splits):

    unique_groups, groups = np.unique(groups, return_inverse=True)
    n_groups = len(unique_groups)

    # Weight groups by their number of occurrences
    n_samples_per_group = np.bincount(groups)

    # Distribute the most frequent groups first
    indices = np.argsort(n_samples_per_group)[::-1]
    n_samples_per_group = n_samples_per_group[indices]

    # Total weight of each fold
    n_samples_per_fold = np.zeros(n_splits)

    # Mapping from group index to fold index
    group_to_fold = np.zeros(len(unique_groups))

    # Distribute samples by adding the largest weight to the lightest fold
    for group_index, weight in enumerate(n_samples_per_group):
        lightest_fold = np.argmin(n_samples_per_fold)
        n_samples_per_fold[lightest_fold] += weight
        group_to_fold[indices[group_index]] = lightest_fold

    indices = group_to_fold[groups]
    print(indices)

# Group 0 with 10 samples, 10 groups with 1 sample
groups = np.concatenate([np.zeros(10), np.arange(1,10)])

_iter_test_indices(groups, 2)
# prints array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1.])
# Two folds with 10 samples each
# First fold contains 1 group, second split contains 10 groups
# Number of samples per fold is balanced, not number of groups per fold
```

Reply:
@amueller Have you had a chance to look at this?